flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapKeyedStateBackend.java

```diff
@@ -259,17 +261,8 @@ private <N, V> StateTable<K, N, V> tryRegisterStateTable(
                                 + ") must not be incompatible with the old state serializer ("
                                 + previousStateSerializer
                                 + ").");
-            }
-
-            // HeapKeyedStateBackend doesn't support ttl state migration currently.
-            if (TtlAwareSerializer.needTtlStateMigration(
-                    previousStateSerializer, newStateSerializer)) {
-                throw new StateMigrationException(
-                        "For heap backends, the new state serializer ("
-                                + newStateSerializer
-                                + ") must not need ttl state migration with the old state serializer ("
-                                + previousStateSerializer
-                                + ").");
+            } else if (stateCompatibility.isCompatibleAfterMigration()) {
+                migrateStateValues(stateDesc, previousStateSerializer, newStateSerializer);
```

Comment:
what happens if the state is not compatible, I suggest we should at least log for that case - or should there be an error in that case?